### DISCUSSION TITLE
How to set background color opacity for div block in insert_htmlbox? It seems blank.

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#3014] [General]
How to set background color opacity for div block in insert_htmlbox? It seems blank.

text = """

 some text<div style="background-color: rgba(0, 255, 255, 0.2);color:gold; font-size: 20px">
    and some text
    with a different background</div>
    and some more text

"""

**Comment by nissansz** [comment]
Now, textwriter morph effect seems perfect.

![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/ecc16b00-7774-4159-beb5-b74775c4c2fd)

**Comment by nissansz** [comment]
textwriter seems faster than insert_htmlbox

**Reply by JorjMcKie** [reply]
It sure is! A lot.
We must not forget what happens in background: HTML/CSS parsing, invocation of the HarfBuzz library for text shaping.
The method has a certain area of purpose and is excellent there.
But if used beyond that it will show limits.

**Comment by nissansz** [comment]
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/934b10dd-2498-42b8-8ecc-f3753ac42515)


Smaller font size, better effect.

**Reply by JorjMcKie** [reply]
Very nice!

**Comment by JorjMcKie** [comment]
See here:
```python
import fitz


rect = fitz.Rect(100, 100, 200, 150)

text = """<div style="background-color:yellow;">Text 1</div>
<div style="background-color:rgb(0,255,0);
text-align:center;color:red;">Text 2</div>"""

doc = fitz.Document()

page = doc.new_page()
page.insert_htmlbox(rect, text)
page.draw_rect(rect, color=(1, 0, 0))
doc.ez_save(__file__.replace(".py", ".pdf"))
```
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/8b3f4a3b-624b-444d-a30b-c7c244d477ee)
So this does not fully fill the rectangle with background color.

There is a way to completely exploit the rectangle's area, by using a crazily large font size like `css="* {font-size: 50px}"`. In that case, the method automatically scales down by an optimized factor and gives you this:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/ef935ca3-11f2-4e0c-a4ef-fb82675a3df3)

**Reply by nissansz** [reply]
text = '<div style="background-color: yellow;color:'+color+'; font-size: '+str(50)+'px">'+text+'<div>'
    page.insert_htmlbox(rect, text, css=css, archive=arch)

even using 50px,  insert_htmlbox does not seem fit.
[3_overlay.pdf](https://github.com/pymupdf/PyMuPDF/files/13901506/3_overlay.pdf)

**Reply by JorjMcKie** [reply]
No, you simply cannot achieve the effect you need in your use case.

**Reply by nissansz** [reply]
It seems TextWriter is more powerful for text effect.

**Reply by JorjMcKie** [reply]
As I said, your use case is not the usual one, and definitely very different from the intention of `insert_htmlbox`.

**Reply by nissansz** [reply]
Understood. Maybe one day, when javascript is supported in insert_htmlbox, it will be very powerful.

**Reply by JorjMcKie** [reply]
What you need here is the opposite: detail influence on the layout. the htmlbox intention is to take away this consideration.

**Reply by nissansz** [reply]
You are pretty correct and expert.

**Comment by JorjMcKie** [comment]
> Understood. Maybe one day, when javascript is supported in insert_htmlbox, it will be very powerful.

It is already very powerful - think of reproduction of images, and HTML tables, and above all correct output of languages requiring text shaping linke Devanagari, Bengali, Tamil, etc.

Textwriter is incapable of doing this.

**Reply by nissansz** [reply]
Will try tables in future, still need time to train table models to recover table structure. Any good method to recover table structure from pdf image?

**Reply by JorjMcKie** [reply]
Not really, just OCR, then `page.find_tables()`.
Problem is that OCR mostly cannot recognize grid lines, which breaks usual table recognition.
So you are dependent on analyzing text positions to sort of guess what **_might_** be a table.

**Reply by JorjMcKie** [reply]
AWS Textract does a good job, but depends on Internet connection, is an AI machinery ... and slow.

**Reply by nissansz** [reply]
Thank you. Will train my own models to test.
